odxtools/statemachine.py

```diff
@@ -0,0 +1,174 @@
+            for stransref in service.state_transition_refs:
+                if stransref.execute(decoded_resp.parameters, decoded_resp_msg.param_dict):
+                   done = True
+                   break
+            if done:
+               break
+    else:
+        raise RuntimeException(f"Cannot execute request for service {service.short_name}")
+    ```
+
+    """
+
+    succeeded: bool
+
+    @property
+    def diag_layer(self) -> "DiagLayer":
+        return self._diag_layer
+
+    @property
+    def state_chart(self) -> StateChart:
+        return self._state_chart
+
+    @property
+    def active_state(self) -> State:
+        return self._active_state
+
+    @active_state.setter
+    def active_state(self, value: State) -> None:
+        self._active_state = value
+
+    def __init__(self, diag_layer: "DiagLayer", state_chart: StateChart) -> None:
+        self.succeeded = True
+        self._diag_layer = diag_layer
+        self._state_chart = state_chart
+        self._active_state = state_chart.start_state
+
+    def execute(self, service: "DiagService", **service_params: Any
+               ) -> Generator[bytes, Union[bytes, bytearray, ParameterValueDict], None]:
+        """Run a diagnostic service and update the state machine
+        depending on the outcome.
+
+        This simplifies error handling, en- and decoding etc.
+
+        Usage example (using asynchronous communication routines):
+
+        ```python
+        ecu = db.ecu_variants.my_ecu_variant
+        fsm = StateMachine(ecu, ecu.state_charts.my_state_chart)
+
+        for raw_request in (executor := fsm.execute(ecu.services.my_service,
+                                                    param1="hello",
+                                                    param2=123)):
+            # send raw request to the ECU (usually using ISO-TP or DoIP), and
+            # receive the response
+            await send_to_ecu(raw_request)
+            while odxtools.is_response_pending(raw_response := await receive_from_ecu()):
+                pass
+
+            executor.send(raw_response)
+
+        print(f"Request send: {fsm.succeeded}")
+        ```
+
+        """
+        if service.request is None:
+            odxraise("Services without requests are not allowed in this context")
+            self.succeeded = False
+            return
+
+        # the service can be executed if any of the specified
+        # precondition states is fulfilled. (TODO: correct?)
+        if service.pre_condition_state_refs is not None:
+            if not any(
+                    x.applies(self, service.request.parameters, service_params)
+                    for x in service.pre_condition_state_refs):
+                # if all preconditions which are applicable are
+                # invalid (i.e., they evaluate to False), we must not
+                # execute the service.
+                self.succeeded = False
```

Comment:
Why not raise?

Reply:
because it is IMO not really an error if the service cannot be run and it would complicate the calling code. (the for loop described in the docstring would need to be made aware of the exception.)